cpp/src/arrow/util/async_generator.h

```diff
@@ -1063,6 +1063,86 @@ AsyncGenerator<T> MakeConcatenatedGenerator(AsyncGenerator<AsyncGenerator<T>> so
   return MergedGenerator<T>(std::move(source), 1);
 }
 
+template <typename T>
+struct Enumerated {
+  util::optional<T> value;
+  int index;
+  bool last;
```

Comment:
Why is `last` useful? One can simply wait for `End()` to arrive, no?

Reply:
The need for this originated from trying to resequence merged streams.  So given a generator that returns streams (A, B, C), each of which return items (1, 2), the merged generator might return something like...

```
A1
B1
A2
C1
C2
B2
```

Any end tokens are removed by the merge operation.  Otherwise the combined stream would terminate after A2.

Reply:
The merged generator _has_ to return an end token at some point, right? I don't understand the concern.

Reply:
Yes, it has to return one end token.  If I return the end token for each child stream I'll end up with 3.  In other words, I want...
```
A1
B1
A2 (last-in-stream)
C1
C2 (last-in-stream)
B2 (last-in-stream)
END
```

...and not...
```
A1
B1
A2
END
C1
C2
END
B2
END
```